### PR TITLE
feat(Combinatorics/SimpleGraph): added matchings in cliques

### PR DESCRIPTION
**Pull request by pimotte**
Prove cliques have matchings exactly when they are even.

In preparation for Tutte's theorem.

---

- [x] depends on: #15652 

[Zulip thread about Tutte](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Tutte's.20theorem)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 1]
```diff
@@ -235,7 +237,60 @@ lemma odd_matches_node_outside {u : Set V} {c : ConnectedComponent (Subgraph.del
     Finset.card_image_of_injective _ (Subtype.val_injective), Set.toFinset_card] at hMeven ⊢
   exact hMeven
 
+
 end Finite
+
+lemma isClique_even_iff_matches [DecidableEq V] (u : Set V) (hu : Set.Finite u)
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:243]:
```suggestion
lemma isClique_even_iff_exists_isMatching [DecidableEq V] (u : Set V) (hu : Set.Finite u)
```
sounds more discoverable

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 2]
```diff
@@ -235,7 +237,60 @@ lemma odd_matches_node_outside {u : Set V} {c : ConnectedComponent (Subgraph.del
...
+  refine ⟨?_ , by
+    rintro ⟨M, ⟨hMl, hMr⟩⟩
+    haveI : Fintype M.verts := hMl ▸ hu.fintype
+    subst hMl
+    simpa [Nat.card_eq_card_finite_toFinset hu, Set.toFinite_toFinset,
+      Set.toFinset_card] using hMr.even_card⟩
+  intro he
+  cases' Set.eq_empty_or_nonempty u with h h
+  · subst h
+    let empty : Subgraph G := {
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:255]:
```suggestion
  obtain rfl | h := u.eq_empty_or_nonempty
  · let empty : Subgraph G := {
```
is a nice pattern

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 3]
```diff
@@ -235,7 +237,60 @@ lemma odd_matches_node_outside {u : Set V} {c : ConnectedComponent (Subgraph.del
...
+      verts := ∅,
+      Adj := fun _ _ ↦ False,
+      adj_sub := False.elim,
+      edge_vert := False.elim }
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:259]: Is that not just `⊥`?
```suggestion
    let empty : Subgraph G := ⊥
```
Then you don't even need a `let`

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 4]
```diff
@@ -235,7 +237,60 @@ lemma odd_matches_node_outside {u : Set V} {c : ConnectedComponent (Subgraph.del
...
+decreasing_by
+· simp_wf
+  refine Set.ncard_lt_ncard ?_ hu
+  exact ⟨Set.diff_subset, by
+    rw [Set.not_subset_iff_exists_mem_not_mem]
+    use x
+    exact ⟨hxy.1, by simp only [Set.mem_diff, Set.mem_insert_iff, Set.mem_singleton_iff, true_or,
+      not_true_eq_false, and_false, not_false_eq_true]⟩⟩
+
+lemma completeGraph_even_iff_matches [Fintype V] [DecidableEq V] :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:289]: Similarly,
```suggestion
lemma completeGraph_even_iff_exists_isMatching [Fintype V] [DecidableEq V] :
```

File: Mathlib/Data/Set/Finite/Parity.lean [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+/-!
+# Finite sets and parity of their cardinality
+
+This file contains lemmas concerning finite sets and parity of their cardinality.
+-/
+
+namespace Set.Finite
+
+variable {V : Type _} {s : Set V}
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Finite/Parity.lean:17]:
```suggestion
variable {V : Type*} {s : Set V}
```

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean (hunk 3 above)
Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Matching.lean]: Ah, that's right! Fixed:)

File: Mathlib/Combinatorics/SimpleGraph/Clique.lean [hunk 6]
```diff
@@ -94,6 +94,10 @@ theorem IsClique.subset (h : t ⊆ s) : G.IsClique s → G.IsClique t := Set.Pai
 theorem isClique_bot_iff : (⊥ : SimpleGraph α).IsClique s ↔ (s : Set α).Subsingleton :=
   Set.pairwise_bot_iff
 
+lemma IsClique.completeGraph : (completeGraph α).IsClique s := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Clique.lean:97]:
```suggestion
protected lemma IsClique.completeGraph : (completeGraph α).IsClique s := by
```

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 7]
```diff
@@ -236,6 +238,52 @@ lemma odd_matches_node_outside {u : Set V} {c : ConnectedComponent (Subgraph.del
   exact hMeven
 
 end Finite
+
+lemma isClique_even_iff_exists_isMatching [DecidableEq V] (u : Set V) (hu : Set.Finite u)
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:242]:
```suggestion
lemma IsClique.even_card_iff_exists_isMatching [DecidableEq V] (u : Set V) (hu : Set.Finite u)
```

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 8]
```diff
@@ -236,6 +238,52 @@ lemma odd_matches_node_outside {u : Set V} {c : ConnectedComponent (Subgraph.del
...
+    (hc : G.IsClique u) : Even (Nat.card u) ↔ ∃ (M : Subgraph G), M.verts = u ∧ M.IsMatching := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:243]: This result is still true if `u` is infinite: `Nat.card u = 0` is even and `u` contains a matching on `u`.
```suggestion
lemma isClique_even_iff_exists_isMatching [DecidableEq V] (u : Set V) (hc : G.IsClique u) :
    Even (Nat.card u) ↔ ∃ (M : Subgraph G), M.verts = u ∧ M.IsMatching := by
```

Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Matching.lean]: I've adapted it to make room for it, but I'm not sure how to get there. The proof strategy for the finite case doesn't work, so I went looking for some sort of lemma about partitioning an infinite set into two infinite sets, but I can't find it. Any thoughts on how to approach this part?

Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Matching.lean]: With some help from @chrisflav and @Raph-DG I have a rough idea how to get there through `Cardinal`. I'll work that out soonish:)

**Comment by pimotte** [comment]
This PR has diverged a little bit too much from main for my taste. I'll withdraw this and work on replacement PRs building on #18878

**Pull request closed without merging**